### PR TITLE
Add elastic net benchmark

### PR DESCRIPTION
**Pull request by dbtsai**

### CODE REVIEW

**Comment by jkbradley** [comment]
@feynmanliang Would you mind taking a look at this when you have a chance?  Thanks!

**Comment by feynmanliang** [comment]
@dbtsai Thanks for the PR. Reviewing now. Do you mind targeting it against `mllib-tests/v1p5`?

File: mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala [hunk 1]
```diff
@@ -138,11 +143,12 @@ class GLMRegressionTest(sc: SparkContext) extends GLMTests(sc) {
...
         s"GLMRegressionTest run with unknown loss ($loss).  Supported values: l2.")
     }
-    if (!Array("none", "l1", "l2").contains(regType)) {
+    if (!Array("none", "l1", "l2", "elastic-net").contains(regType)) {
       throw new IllegalArgumentException(
-        s"GLMRegressionTest run with unknown regType ($regType).  Supported values: none, l1, l2.")
+        s"GLMRegressionTest run with unknown regType ($regType).  Supported values: none, l1, l2, elastic-net")
     }
-    if (!Array("sgd").contains(optimizer)) { // only SGD supported in Spark 1.1
+    if (!(Array("sgd").contains(optimizer) || Array("elastic-net").contains(regType))) {
```
Comment (feynmanliang) [inline mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: Does this change break perf-testing `regType` = {none,l1,l2}?

File: mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala [hunk 2]
```diff
@@ -138,11 +143,12 @@ class GLMRegressionTest(sc: SparkContext) extends GLMTests(sc) {
...
+      // only SGD supported in Spark 1.4 or elastic-net supported through OWLQN/LBFGS
```
Comment (feynmanliang) [inline mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: Update doc to Spark 1.5 after changing this PR against `v1p5`

File: mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala [hunk 3]
```diff
@@ -75,11 +78,12 @@ abstract class RegressionAndClassificationTests[M](sc: SparkContext) extends Per
...
-  val NUM_ITERATIONS = ("num-iterations",   "number of iterations for the algorithm")
-  val REG_TYPE =       ("reg-type",   "type of regularization: none, l1, l2")
-  val REG_PARAM =      ("reg-param",   "the regularization parameter against overfitting")
-  val OPTIMIZER =      ("optimizer", "optimization algorithm: sgd, lbfgs")
+  val STEP_SIZE =         ("step-size",   "step size for SGD")
+  val NUM_ITERATIONS =    ("num-iterations",   "number of iterations for the algorithm")
+  val REG_TYPE =          ("reg-type",   "type of regularization: none, l1, l2, elastic-net")
+  val ELASTIC_NET_PARAM = ("elastic-net-Param",   "elastic-net param, 0.0 for L2, and 1.0 for L1")
+  val REG_PARAM =         ("reg-param",   "the regularization parameter against overfitting")
+  val OPTIMIZER =         ("optimizer", "optimization algorithm: sgd, lbfgs")
```
Comment (feynmanliang) [inline mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: We should probably remove `lbfgs` from here if it's not supported in the perf tests yet

File: mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala (hunk 1 above)
Comment (feynmanliang) [inline mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: This breaks the previous check i.e. now I can set optimizer="bla" and regType="elastic-net" and things will work.

What about refactoring 146-154 to check the combination of optimizer and regularization is either (sgd, {none, l1, l2}) or (lbfgs, elastic-net) and reporting error appropriately (e.g. sgd can only be used with regTypes ..., lbfgs can only be used with elastic-net regularization)

File: mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala [hunk 4]
```diff
@@ -138,11 +143,12 @@ class GLMRegressionTest(sc: SparkContext) extends GLMTests(sc) {
...
+    if (!Array("none", "l1", "l2", "elastic-net").contains(regType)) {
       throw new IllegalArgumentException(
-        s"GLMRegressionTest run with unknown regType ($regType).  Supported values: none, l1, l2.")
+        s"GLMRegressionTest run with unknown regType ($regType).  Supported values: none, l1, l2, elastic-net")
     }
-    if (!Array("sgd").contains(optimizer)) { // only SGD supported in Spark 1.1
+    if (!(Array("sgd").contains(optimizer) || Array("elastic-net").contains(regType))) {
+      // only SGD supported in Spark 1.4 or elastic-net supported through OWLQN/LBFGS
       throw new IllegalArgumentException(
         s"GLMRegressionTest run with unknown optimizer ($optimizer). Supported values: sgd.")
```
Comment (feynmanliang) [inline mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: Technically this is no longer true (since `lbfgs` is used for elasticnet)

File: mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala [hunk 5]
```diff
@@ -160,6 +166,18 @@ class GLMRegressionTest(sc: SparkContext) extends GLMTests(sc) {
         val rr = new RidgeRegressionWithSGD().setIntercept(addIntercept = true)
         rr.optimizer.setNumIterations(numIterations).setStepSize(stepSize).setRegParam(regParam)
         rr.run(rdd)
+      case ("l2", "elastic-net") =>
+        println("WARNING: Liner Regression with elastic-net in ML package ignores numIterations, " +
+          "stepSize in this Spark version.")
```
Comment (feynmanliang) [inline mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: State explicit spark version

File: mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala [hunk 6]
```diff
@@ -216,11 +235,11 @@ class GLMClassificationTest(sc: SparkContext) extends GLMTests(sc) {
...
         s"GLMClassificationTest run with unknown loss ($loss).  Supported values: logistic, hinge.")
     }
-    if (!Array("none", "l1", "l2").contains(regType)) {
+    if (!Array("none", "l1", "l2", "elastic-net").contains(regType)) {
       throw new IllegalArgumentException(s"GLMClassificationTest run with unknown regType" +
-        s" ($regType).  Supported values: none, l1, l2.")
+        s" ($regType).  Supported values: none, l1, l2, elastic-net.")
     }
-    if (!Array("sgd", "lbfgs").contains(optimizer)) {
+    if (!(Array("sgd", "lbfgs").contains(optimizer) || Array("elastic-net").contains(regType))) {
```
Comment (feynmanliang) [inline mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: Same feedback as L146-154

File: mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala [hunk 7]
```diff
@@ -232,6 +251,14 @@ class GLMClassificationTest(sc: SparkContext) extends GLMTests(sc) {
         println("WARNING: LogisticRegressionWithLBFGS ignores numIterations, stepSize" +
           " in this Spark version.")
         new LogisticRegressionWithLBFGS().run(rdd)
+      case ("logistic", "elastic-net", _) =>
+        println("WARNING: Logistic Regression with elastic-net in ML package ignores numIterations, " +
+          "stepSize in this Spark version.")
```
Comment (feynmanliang) [inline mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: State spark version

**Comment by jkbradley** [comment]
Pinging for updates.  Thanks!

**Comment by dbtsai** [comment]
Gonna work on this tonight.

File: mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala [hunk 8]
```diff
@@ -160,6 +166,18 @@ class GLMRegressionTest(sc: SparkContext) extends GLMTests(sc) {
         val rr = new RidgeRegressionWithSGD().setIntercept(addIntercept = true)
         rr.optimizer.setNumIterations(numIterations).setStepSize(stepSize).setRegParam(regParam)
         rr.run(rdd)
+      case ("l2", "elastic-net") =>
+        println("WARNING: Liner Regression with elastic-net in ML package ignores numIterations, " +
+          "stepSize in this Spark version.")
+        val rr = new LinearRegression().setElasticNetParam(elasticNetParam).setRegParam(regParam)
+        val sqlContext = new SQLContext(rdd.context)
+        import sqlContext.implicits._
+        val mlModel = rr.fit(rdd.toDF())
```
Comment (feynmanliang) [inline mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: Note: generating the DF inside of `runTests` is OK for now, but it introduces the DF generation overhead into the performance metrics for ML pipeline algorithms. We will eventually need to refactor spark-perf to use DataFrames.

File: mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala (hunk 3 above)
Comment (dbtsai) [inline mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: elasticnet is using `lbfgs/owlqn`, but lbfgs is also used in old mllib LoR. I will probably document it.

File: mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala (hunk 1 above)
Comment (dbtsai) [inline mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: Sounds good.

File: mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala (hunk 8 above)
Comment (dbtsai) [inline mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: Although this requires major refactoring, but I think in the long term, we should go data frame.

File: mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala (hunk 5 above)
Comment (dbtsai) [inline mllib-tests/v1p4/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: I modified it in the code. We may want to refactor it to use some static version string instead of hard coding the version.

File: mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala [hunk 9]
```diff
@@ -80,11 +81,12 @@ abstract class RegressionAndClassificationTests[M](sc: SparkContext) extends Per
...
 
-  val STEP_SIZE =      ("step-size",   "step size for SGD")
-  val NUM_ITERATIONS = ("num-iterations",   "number of iterations for the algorithm")
-  val REG_TYPE =       ("reg-type",   "type of regularization: none, l1, l2")
-  val REG_PARAM =      ("reg-param",   "the regularization parameter against overfitting")
-  val OPTIMIZER =      ("optimizer", "optimization algorithm: sgd, lbfgs")
+  val STEP_SIZE =         ("step-size",   "step size for SGD")
+  val NUM_ITERATIONS =    ("num-iterations",   "number of iterations for the algorithm")
+  val REG_TYPE =          ("reg-type",   "type of regularization: none, l1, l2, elastic-net")
+  val ELASTIC_NET_PARAM = ("elastic-net-Param",   "elastic-net param, 0.0 for L2, and 1.0 for L1")
```
Comment (jkbradley) [inline mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: lower-case "Param" in param name

File: mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala [hunk 10]
```diff
@@ -165,6 +176,18 @@ class GLMRegressionTest(sc: SparkContext) extends GLMTests(sc) {
         val rr = new RidgeRegressionWithSGD().setIntercept(addIntercept = true)
         rr.optimizer.setNumIterations(numIterations).setStepSize(stepSize).setRegParam(regParam)
         rr.run(rdd)
+      case ("l2", "elastic-net") =>
+        println("WARNING: Liner Regression with elastic-net in ML package ignores numIterations, " +
```
Comment (jkbradley) [inline mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: It should use numIterations (for maxIter).

typo: "Linear"

File: mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala [hunk 11]
```diff
@@ -237,6 +273,14 @@ class GLMClassificationTest(sc: SparkContext) extends GLMTests(sc) {
         println("WARNING: LogisticRegressionWithLBFGS ignores numIterations, stepSize" +
           " in this Spark version.")
         new LogisticRegressionWithLBFGS().run(rdd)
+      case ("logistic", "elastic-net", _) =>
+        println("WARNING: Logistic Regression with elastic-net in ML package ignores numIterations, " +
```
Comment (jkbradley) [inline mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: set maxIter

**Comment by jkbradley** [comment]
@dbtsai Have you run this yet?  If so, could you please add reasonable default settings to the config.py.template file?  Thanks!

**Comment by dbtsai** [comment]
@jkbradley Do I need to add "elastic-net-param" somewhere other than MLAlgorithmTests.scala? I keep getting

```java
Java options: -Dspark.storage.memoryFraction=0.66 -Dspark.serializer=org.apache.spark.serializer.JavaSerializer -Dspark.locality.wait=60000000 -Dspark.shuffle.manager=SORT
Options: glm-regression --num-trials=10 --inter-trial-wait=3 --num-partitions=1 --random-seed=5 --num-examples=1000 --num-features=10000 --loss=l2 --num-iterations=20 --optimizer=lbfgs --reg-type=elastic-net --elastic-net-param=0.0 --reg-param=0.0

15/08/18 23:59:40 INFO SparkDeploySchedulerBackend: SchedulerBackend is ready for scheduling beginning after reached minRegisteredResourcesRatio: 0.0
Exception in thread "main" joptsimple.UnrecognizedOptionException: 'elastic-net-param' is not a recognized option
	at joptsimple.OptionException.unrecognizedOption(OptionException.java:89)
	at joptsimple.OptionParser.handleLongOptionToken(OptionParser.java:429)
	at joptsimple.OptionParserState$2.handleArgument(OptionParserState.java:56)
	at joptsimple.OptionParser.parse(OptionParser.java:361)
	at mllib.perf.PerfTest.initialize(PerfTest.scala:21)
	at mllib.perf.TestRunner$.main(TestRunner.scala:57)
	at mllib.perf.TestRunner.main(TestRunner.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:672)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:205)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:120)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
```

File: mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala [hunk 12]
```diff
@@ -80,11 +81,12 @@ abstract class RegressionAndClassificationTests[M](sc: SparkContext) extends Per
...
-  val OPTIMIZER =      ("optimizer", "optimization algorithm: sgd, lbfgs")
+  val STEP_SIZE =         ("step-size",   "step size for SGD")
+  val NUM_ITERATIONS =    ("num-iterations",   "number of iterations for the algorithm")
+  val REG_TYPE =          ("reg-type",   "type of regularization: none, l1, l2, elastic-net")
+  val ELASTIC_NET_PARAM = ("elastic-net-param",   "elastic-net param, 0.0 for L2, and 1.0 for L1")
+  val REG_PARAM =         ("reg-param",   "the regularization parameter against overfitting")
+  val OPTIMIZER =         ("optimizer", "optimization algorithm (elastic-net only supports lbfgs): sgd, lbfgs")
 
   intOptions = intOptions ++ Seq(NUM_ITERATIONS)
   doubleOptions = doubleOptions ++ Seq(STEP_SIZE, REG_PARAM)
```
Comment (feynmanliang) [inline mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: Add `ELASTIC_NET_PARAM` to `doubleOptions`

File: config/config.py.template [hunk 13]
```diff
@@ -429,6 +429,26 @@ MLLIB_GLM_REGRESSION_TEST_OPTS = MLLIB_GLM_TEST_OPTS + [
 MLLIB_TESTS += [("glm-regression", MLLIB_PERF_TEST_RUNNER, SCALE_FACTOR,
     MLLIB_JAVA_OPTS, [ConstantOption("glm-regression")] + MLLIB_GLM_REGRESSION_TEST_OPTS)]
 
+if MLLIB_SPARK_VERSION >= 1.5:
+    MLLIB_GLM_ELASTIC_NET_REGRESSION_TEST_OPTS = MLLIB_REGRESSION_CLASSIFICATION_TEST_OPTS + [
```
Comment (feynmanliang) [inline config/config.py.template]: I am getting

```
Missing required option(s) ['step-size', 'intercept', 'epsilon']
```

when I run this, so I think you will need to add those options here

**Comment by dbtsai** [comment]
@feynmanliang Thanks. It's working now. I am going to run it in our cluster on Monday.

**Comment by dbtsai** [comment]
Ping @feynmanliang  @jkbradley

File: config/config.py.template [hunk 14]
```diff
@@ -429,6 +429,32 @@ MLLIB_GLM_REGRESSION_TEST_OPTS = MLLIB_GLM_TEST_OPTS + [
...
+        OptionSet("loss", ["l2"]),
+        # The max number of iterations for LBFGS/OWLQN
+        OptionSet("num-iterations", [20]),
+        # LBFGS/OWLQN is used with elastic-net regularization.
+        OptionSet("optimizer", ["lbfgs"]),
+        # Using elastic-net regularization.
+        OptionSet("reg-type", ["elastic-net"]),
+        # Runs with L2 (param = 0.0), L1 (param = 1.0), and mixing L1/L2.
+        OptionSet("elastic-net-param", [0.0, 0.5, 1.0]),
+        # Runs with lambda = [0.0, 0.5]
```
Comment (feynmanliang) [inline config/config.py.template]: nit: I would prefer if this was just 0.5 to keep perf-test parameter space small

File: config/config.py.template [hunk 15]
```diff
@@ -429,6 +429,32 @@ MLLIB_GLM_REGRESSION_TEST_OPTS = MLLIB_GLM_TEST_OPTS + [
...
+        # Loss to minimize: l2 (squared error)
+        OptionSet("loss", ["l2"]),
+        # The max number of iterations for LBFGS/OWLQN
+        OptionSet("num-iterations", [20]),
+        # LBFGS/OWLQN is used with elastic-net regularization.
+        OptionSet("optimizer", ["lbfgs"]),
+        # Using elastic-net regularization.
+        OptionSet("reg-type", ["elastic-net"]),
+        # Runs with L2 (param = 0.0), L1 (param = 1.0), and mixing L1/L2.
+        OptionSet("elastic-net-param", [0.0, 0.5, 1.0]),
```
Comment (feynmanliang) [inline config/config.py.template]: Did you see significant changes in performance from varying these? If not, I would just stick with 0.5

File: config/config.py.template [hunk 16]
```diff
@@ -447,6 +473,30 @@ MLLIB_TESTS += [("glm-classification", MLLIB_PERF_TEST_RUNNER, SCALE_FACTOR,
...
+if MLLIB_SPARK_VERSION >= 1.5:
+    MLLIB_GLM_ELASTIC_NET_CLASSIFICATION_TEST_OPTS = MLLIB_GLM_CLASSIFICATION_TEST_OPTS + [
+        # In GLM classification with elastic-net regularization, only logistic loss is supported.
+        OptionSet("loss", ["logistic"]),
+        # LBFGS/OWLQN is used with elastic-net regularization.
+        OptionSet("optimizer", ["lbfgs"]),
+        # Using elastic-net regularization.
+        OptionSet("reg-type", ["elastic-net"]),
+        # Runs with L2 (param = 0.0), L1 (param = 1.0), and mixing L1/L2.
+        OptionSet("elastic-net-param", [0.0, 0.5, 1.0]),
```
Comment (feynmanliang) [inline config/config.py.template]: ditto with above comments

File: config/config.py.template [hunk 17]
```diff
@@ -455,7 +505,7 @@ NAIVE_BAYES_TEST_OPTS = MLLIB_REGRESSION_CLASSIFICATION_TEST_OPTS + [
     # Naive Bayes smoothing lambda.
     OptionSet("nb-lambda", [1.0]),
     # Model type: either Multinomial or Bernoulli
-    OptionSet("model-type", ["Multinomial"]),
```
Comment (feynmanliang) [inline config/config.py.template]: While you're at it, can you revert the change on this line since it will cause a conflict with #85

File: mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala [hunk 18]
```diff
@@ -214,18 +237,31 @@ class GLMClassificationTest(sc: SparkContext) extends GLMTests(sc) {
...
-    }
-    if (!Array("sgd", "lbfgs").contains(optimizer)) {
+
+    if (Array("sgd").contains(optimizer)) {
+      if (!Array("none", "l1", "l2").contains(regType)) {
+        throw new IllegalArgumentException(
+          s"GLMRegressionTest run with unknown regType ($regType) with sgd.  Supported values: none, l1, l2.")
+      }
+    } else if (Array("lbfgs").contains(optimizer)) {
+      if (Array("hinge").contains(loss)) {
```
Comment (feynmanliang) [inline mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: I would check `!Array("logistic")` instead to guard against `optimizer=lbfgs, loss=some_nonsense`

Comment (feynmanliang) [inline mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: nit: I think L260-263 is sufficient and this can be cut

**Comment by feynmanliang** [comment]
Made a pass, had some minor comments

File: mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala (hunk 18 above)
Comment (dbtsai) [inline mllib-tests/v1p5/src/main/scala/mllib/perf/MLAlgorithmTests.scala]: I think checking the loss is still needed. Otherwise, hinge -> lbfgs -> elastic-net will pass.

**Comment by jkbradley** [comment]
I'll go ahead and merge this.  Thanks!
